proposals/0167-loader-v4.md

```diff
@@ -0,0 +1,477 @@
+choose the user having to announce the length of the program to be uploaded
+explicitly before the actual upload starts.
+
+## New Terminology
+
+The _current slot_ is as in the Clock sysvar.
+
+Delay visibility: The changed version of a program only becomes available after
+the current slot ends. Thus, the first transaction in the next slot can invoke
+it.
+
+Deployment cooldown: There can be at most one deployment per program in the
+same slot. Subsequent deployments have to wait for the next slot.
+
+## Detailed Design
+
+The feature gate must enable loader-v4 program management and execution.
+
+### Program Account Layout
+
+Accounts of programs owned by loader-v4 must have the following layout:
+
+- Header (which is 48 bytes long):
+  - `u32` status enum:
+    - Enum variant `0u64`: `Uninitalized`, account was zero-filled externally
+    - Enum variant `1u64`: `NeverBeenDeployed`, used as write buffer
+    - Enum variant `2u64`: `Retracted`, program is in maintenance
+    - Enum variant `3u64`: `Deployed`, program is ready to be executed
+    - Enum variant `4u64`: `Finalized`, same as `Deployed`, but can not be
+    modified anymore
+  - `u32` Length of the executable in bytes (see "Body" of the account)
+  - `u64` Slot in which the program was last deployed.
+  - `[u8; 32]` Authority address which can send program management
+  instructions.
+- Body:
+  - `[u8]` The programs executable file
+
+### Program Account Header Verification
+
+Verification the program account checks in the following order that:
+
+- the owner of the program account is loader-v4,
+otherwise throw `InvalidAccountOwner`
+- the program account is at least as long enough for the header (48 bytes),
+otherwise throw `AccountDataTooSmall`
+- the program account is writable, otherwise throw `InvalidArgument`
+- the provided authority (instruction account at index 1) signed,
+otherwise throw `MissingRequiredSignature`
+- the authority stored in the program account is the one provided,
+otherwise throw `IncorrectAuthority`
+- the status stored in the program account is not `Uninitalized`,
+otherwise throw `InvalidArgument`
+- the status stored in the program account is not `Finalized`,
+otherwise throw `Immutable`
+
+### Execution / Invocation
+
+Invoking programs owned by loader-v4 checks in the following order that:
+
+- the owner of the program account is loader-v4
+- the program account is at least as long enough for the header
+- the status stored in the program account is `Deployed` or `Finalized`
+- the program account was not deployed within the current slot (delay
+visibility)
+- the executable file stored in the program account passes executable
+verification. The executable file starts after the 48 byte header and has the
+length defined in the header, which could be less than the account length.
+
+failing any of the above checks must throw `UnsupportedProgramId`.
+
+### Program Management Instructions
```

Comment:
Should also add a note saying that instruction data length is limited to 1232 bytes to match loader-v3 behavior